File: GILua/src/main/java/org/anime_game_servers/gi_lua/models/scene/group/SceneTrigger.java
package org.anime_game_servers.gi_lua.models.scene.group;

import lombok.Data;
import lombok.NoArgsConstructor;
import org.anime_game_servers.core.base.annotations.lua.LuaNames;
import org.anime_game_servers.gi_lua.models.scene.SceneMeta;

@Data
@NoArgsConstructor
public class SceneTrigger {
    /**
     * When the trigger count is set to this, it only gets unregistered on a return of != 0 from the trigger function.
     */
    public static final int INF_TRIGGERS = 0;

    private String name;
    @LuaNames("config_id")
    private int config_id;
    private int event;
    @LuaNames("trigger_count")
    private int trigger_count = 1;
    private String source;
    private String condition;
    private String action;
    private String tag;
    @LuaNames("tlog_tag")
    private String tlog_tag;
    @LuaNames("forbid_guest")
    private boolean forbid_guest = true;

    private transient int groupId;
    private transient int blockId;
    private transient SceneMeta sceneMeta;
}

File: GILua/src/main/java/org/anime_game_servers/gi_lua/models/scene/group/SceneTrigger.kt
package org.anime_game_servers.gi_lua.models.scene.group

import org.anime_game_servers.core.base.annotations.lua.LuaNames
import org.anime_game_servers.core.base.interfaces.IntKey
import org.anime_game_servers.gi_lua.models.scene.SceneMeta

data class SceneTrigger (
    val name: String? = null,

    @field:LuaNames("config_id")
    val configId: Int = 0,
    val event: Int = 0,

    @field:LuaNames("trigger_count")
    val triggerCount: Int = 1,
    val source: String? = null,
    val condition: String? = null,
    val action: String? = null,
    val tag: String? = null,

    @field:LuaNames("tlog_tag")
    val tlogTag: String? = null,

    @field:LuaNames("forbid_guest")
    val forbidGuest: Boolean = true,

    // not directly part of the lua table
    @Transient
    var groupId: Int = 0,

    @Transient
    var blockId: Int = 0,

    @Transient
    var sceneMeta: SceneMeta? = null,
): IntKey {

    override fun getIntKey() = configId

    companion object {
        /**
         * When the trigger count is set to this, it only gets unregistered on a return of != 0 from the trigger function.
         */
        const val INF_TRIGGERS: Int = 0
    }
}
